### ISSUE TITLE
Choose a library for generating charts in the portal

### ISSUE THREAD

**Issue by chrisguindon** [EclipseFdn/react-eclipsefdn-members#169]
We are currently planning to use https://www.chartjs.org/ as our open-source JavaScript library of choice for data visualization. Charts.js  supports 8 chart types: bar, line, area, pie, bubble, radar, polar, and scatter.

We are also considering Apache ECharts: https://echarts.apache.org/en/index.html

Let's use this bug to discuss the pros and cons.

**Comment by linkfang** [comment]
Echart
Pros:
- Overall, more chart types. It has all 3 charts type used in design. 
Area stacked chart, combo line/bar chart, and map/GEO chart.

Cons:
- Some of the example code is using Chinese for labels (such as, weekdays, countries, etc.) and comments.

Charts.js
Pros:
- More popular than Echart (1.5 million weekly downloads VS 0.3 million)
- Pure English docs and codes

Cons:
- Overall, less chart types. Doesn't have built-in area stacked chart and map/GEO chart, need to install extra libraries.

I personally prefer Echart, but if the Chinese label is a big impact to the team, I am good with Charts.js as well.

I am doing some demo charts using these 2, will update some thoughts after that.

**Comment by linkfang** [comment]
Here is the preview of the charts created with both libraries:
https://deploy-preview-189--eclipsefdn-react-members.netlify.app/portal#committers-contributors

Here are some of my thoughts and feelings about those 2.

- Echarts is more interactive
For example, the tooltip in bar line chart. 
When user hovers the column in the chart, he/she will see the tooltip. And it will follow the cursor. 
For Chartjs, user has to hover on the bar or the point on the line to see the tooltip. 
As a user, I prefer Echart way of presenting the tooltip, as I don't need to try to follow both the cursor and the tooltip when I move my mouse.

- Echarts has more chart types
For example, the area stacked chart. 
Echart has the curve version which align with our design, but Chartjs only has the standard version. 
Same for the geo/map chart. Need to do some extra work (installing another library) with Chartjs to make it happen.

- Echarts supports more features
Such as dark mode and exporting the chart as an image. We don't need to add extra code to make these happen.

- Chartjs has a more fluent responsive design
When resize the browser window, Chartjs feels much more fluent. It looks like Echart re-renders/re-draw in a much lower frequency.

- Chartjs has a full English doc while Echart uses some Chinese characters for labels (country name, months ect.)

- Both libraries are straightforward
Basically, we just need to offer the data and or the option and apply it/them to the chart component and that's it. 
Need to take care of the data/option to change the chart style or behaviour. 

In conclusion, I personally prefer Echarts (or say strongly recommend it), but as I said, if the Chinese labels could be a big impact to the team, I am good with Chartjs as well.

**Comment by autumnfound** [comment]
> Here is the preview of the charts created with both libraries:
> https://deploy-preview-189--eclipsefdn-react-members.netlify.app/portal#committers-contributors
> 
> Here are some of my thoughts and feelings about those 2.
> 
> * Echarts is more interactive
>   For example, the tooltip in bar line chart.
>   When user hovers the column in the chart, he/she will see the tooltip. And it will follow the cursor.
>   For Chartjs, user has to hover on the bar or the point on the line to see the tooltip.
>   As a user, I prefer Echart way of presenting the tooltip, as I don't need to try to follow both the cursor and the tooltip when I move my mouse.
> * Echarts has more chart types
>   For example, the area stacked chart.
>   Echart has the curve version which align with our design, but Chartjs only has the standard version.
>   Same for the geo/map chart. Need to do some extra work (installing another library) with Chartjs to make it happen.
> * Echarts supports more features
>   Such as dark mode and exporting the chart as an image. We don't need to add extra code to make these happen.
> * Chartjs has a more fluent responsive design
>   When resize the browser window, Chartjs feels much more fluent. It looks like Echart re-renders/re-draw in a much lower frequency.
> * Chartjs has a full English doc while Echart uses some Chinese characters for labels (country name, months ect.)
> * Both libraries are straightforward
>   Basically, we just need to offer the data and or the option and apply it/them to the chart component and that's it.
>   Need to take care of the data/option to change the chart style or behaviour.
> 
> In conclusion, I personally prefer Echarts (or say strongly recommend it), but as I said, if the Chinese labels could be a big impact to the team, I am good with Chartjs as well.

In a world where these 2 plugins were equal outside of the functionality, then Echarts would be a very compelling product. The problem is that ChartJS is massively more popular (1.5m vs 300k DL [source](https://www.npmtrends.com/chart.js-vs-echarts)) which brings forward a lot of different discussions.

When adding a product to a stack, we need to consider community support + reactivity of the dev team. If there are issues/questions had, it is much easier to find answers with ChartJS than echarts, even before we take into account the language demographic. With the language demographic, it makes it harder to recommend echarts.

Another thing we need to take into consideration as I hinted before is language demographics of support and documentation. While English is present for Echarts, it seems incomplete in comparison to the Chinese documentation. As English is the common language, we need to be sure that the support for that language is what is taken into account when looking at products.

An additional thing to consider is the size of the plugin itself minzipped. ChartJS is 61kb, where echarts is a whopping 300kb, which is the size of all of our vendor and first-party JS libs combined. That is a huge thing to consider when we are looking to implement new features and is something that you can't leave out of a discussion.

**Comment by chrisguindon** [comment]
I look at all the pros and cons that you both did post. From a look and feel perspective, I do prefer echarts and I love that's it a swiss army knife solution when it comes to charting. However, the tradeoff are too important to ignore (size of javascript code + documentation)

192 kb: no chart
259.34 KB  : with chart js only
518.01 KB: with echart only

I decided to choose chart.js because echarts will increase the size of our javascript by more than 100%. I also needed to consider that English is not the primary language for their documentation. I think it makes sense for us to standardize on a solution that has a small footprint as I could see us using this solution in future projects.
